### PR TITLE
Issue #67: Backfilled metric - online users

### PR DESCRIPTION
**Pull request by marcghaly**
Hi @jaypha, hi @brendanheywood, 

This is PR for issue #67.

Regards,

Marc

### CODE REVIEW

File: classes/metric/concurrent_users_metrics.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_cloudmetrics\metric;
+
+use cltr_database\collector;
+
+/**
+ * Metric class for new users.
```
Comment (jaypha) [inline classes/metric/concurrent_users_metrics.php]: Need to describe this class.

File: classes/metric/active_users_metric.php [hunk 2]
```diff
@@ -55,4 +55,13 @@ public function get_metric_item(): metric_item {
         $users = $DB->count_records_select('user', 'lastlogin > ?', [$now - $this->get_time_window()]);
         return new metric_item($this->get_name(), $now, $users, $this);
     }
+
+    /**
+     * Metric's ability to be backfilled.
+     *
+     * @return bool
+     */
+    public function is_backfillable(): bool {
```
Comment (jaypha) [inline classes/metric/active_users_metric.php]: If this is going to appear in every metric, then it should also be put into the base class, even if it is abstract. Although it should probably default to returning false.

File: classes/metric/concurrent_users_metrics.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * Returns records for backfilled metric.
+     *
+     * @param array $data Array of data specifying timeframe and sample to backfill.
+     * @return array|null
+     */
+    public function get_backfilled_metric(array $data): ?array {
+        global $DB;
+        $collector = new collector();
```
Comment (jaypha) [inline classes/metric/concurrent_users_metrics.php]: Metrics are not supposed to be aware of collectors. Especially specific ones. You should be able to match any collector to any metric and have it work.

File: classes/metric/concurrent_users_metrics.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        return true;
+    }
+
+    /**
+     * Returns records for backfilled metric.
+     *
+     * @param array $data Array of data specifying timeframe and sample to backfill.
+     * @return array|null
+     */
+    public function get_backfilled_metric(array $data): ?array {
```
Comment (jaypha) [inline classes/metric/concurrent_users_metrics.php]: This should also be declared in the base class, with a default of doing nothing.

File: collector/database/classes/collector.php [hunk 5]
```diff
@@ -39,19 +45,41 @@ public function record_metric(metric_item $item) {
...
      * @throws \dml_exception
      */
-    public function get_metrics($metricnames = null, $since = false): array {
+    public function get_metrics($metricnames = null, $metricclass = null, $since = false, $data = []): array {
         global $DB;
         $starting = '';
+
+        if (!is_null($metricclass)) {
+            if ($metricclass->is_backfillable()) {
+                $backfilleddata = $metricclass->get_backfilled_metric($data);
```
Comment (jaypha) [inline collector/database/classes/collector.php]: Is this the best place for it?
This function (which I now realise is poorly named. It should be called 'get_saved_metric_data', or something similar), is about retrieving saved metric data, not about generating them. Perhaps this should be done where metric data is originally generated.

This function is unique to cltr_database\collector. Other collectors may have the ability to record backfill data, but not have the ability to retrieve it (like AWS).

File: collector/database/classes/form/metric_backfill_form.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Form for specifying user metric data to be backfilled.
+ *
+ * @package    cltr_database
+ * @copyright  2022 Catalyst IT Australia Pty Ltd
+ * @author     Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
```
Comment (jaypha) [inline collector/database/classes/form/metric_backfill_form.php]: Excess whitespace.

**Comment by jaypha** [comment]
There are some issues here which I have concerns about. I'll be happy to discuss if you want to.

File: classes/metric/concurrent_users_metrics.php [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * Metric class for new users.
+ *
+ * @package    metric_foobar
+ * @author     Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @copyright  2022, Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class concurrent_users_metrics extends builtin {
```
Comment (jaypha) [inline classes/metric/concurrent_users_metrics.php]: Isn't this the same as online_users_metric?

File: classes/metric/concurrent_users_metrics.php [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    /**
+     * Retrieves the metric.
+     *
+     * @param  $backfilledrecord If data is backfilled, record is passed here.
+     * @return metric_item
+     */
+    public function get_metric_item($record = null): metric_item {
+        return new metric_item($this->get_name(), $record->time, $record->value, $this);
```
Comment (jaypha) [inline classes/metric/concurrent_users_metrics.php]: How is this supposed to work during normal metric gathering?
What happens when $record is null?

File: classes/metric/concurrent_users_metrics.php (hunk 1 above)
Comment (marcghaly) [reply classes/metric/concurrent_users_metrics.php]: Fixed

File: classes/metric/active_users_metric.php (hunk 2 above)
Comment (marcghaly) [reply classes/metric/active_users_metric.php]: Fixed

File: classes/metric/concurrent_users_metrics.php (hunk 3 above)
Comment (marcghaly) [reply classes/metric/concurrent_users_metrics.php]: I have split and renamed this function, please let me know if this suits.

File: classes/metric/concurrent_users_metrics.php (hunk 4 above)
Comment (marcghaly) [reply classes/metric/concurrent_users_metrics.php]: Fixed

File: collector/database/classes/collector.php (hunk 5 above)
Comment (marcghaly) [reply collector/database/classes/collector.php]: I have split this function with others - as stated in #67, I also added a declaration whether or not a collector supports metric recording - should there also be a declaration as whether they can be retrieved or not ?

File: collector/database/classes/form/metric_backfill_form.php (hunk 6 above)
Comment (marcghaly) [reply collector/database/classes/form/metric_backfill_form.php]: Fixed

File: classes/metric/concurrent_users_metrics.php (hunk 7 above)
Comment (marcghaly) [reply classes/metric/concurrent_users_metrics.php]: I understood that online users are users in the sense of right now whereas concurrent users are values from the past, if those are the same I will remove the extra one.

File: classes/metric/concurrent_users_metrics.php (hunk 8 above)
Comment (marcghaly) [reply classes/metric/concurrent_users_metrics.php]: If possible I would like to discuss this in a bbb maybe your Monday or Tuesday morning (Sunday or Monday evening for me) ?

Comment (jaypha) [reply classes/metric/concurrent_users_metrics.php]: Tues morning (Aussie time). If you can specify a specific time, I can make sure I am here for it. Any time is fine except for the 10:00 meeting.

Comment (jaypha) [reply classes/metric/concurrent_users_metrics.php]: That is, I have meetings between 10:00 and 10:30 (Sydney time). Any other time is fine.

File: classes/metric/concurrent_users_metrics.php (hunk 7 above)
Comment (jaypha) [reply classes/metric/concurrent_users_metrics.php]: In this case the online_users_metric is redundant as concurrent_users_metric can fill both roles.

File: collector/database/classes/collector.php [hunk 9]
```diff
@@ -72,4 +89,33 @@ public function get_metrics($metricnames = null, $since = false): array {
...
         return $DB->get_records_sql($sql, $params);
     }
+
+    /**
+     * Records retrieved data in collector.
+     *
+     * @param  object $metricclass Class representing metric.
+     * @param  array $data Array of metric items.
+     */
+    public function record_saved_metrics(object $metricclass, array $metricitems = []) {
```
Comment (jaypha) [inline collector/database/classes/collector.php]: 'object' as a type hint is not supported in older versions of PHP. It should be '\tool_cloudmetrics\metric\base' instead.

**Comment by jaypha** [comment]
Hi Marc.

I don't think a meeting is necessary. You have addressed my primary concerns for now. I'm a little iffy about having it controlled from within the database collector, but that's something I'd rather leave to Brendan.

Just that one fix causing the CI problem, and then I am OK to merge. Get Brendan to have a look at the UI aspect.

Regards

File: classes/metric/concurrent_users_metrics.php (hunk 7 above)
Comment (brendanheywood) [reply classes/metric/concurrent_users_metrics.php]: +1, this should just be back filling the online_users_metric

File: classes/metric/concurrent_users_metrics.php (hunk 8 above)
Comment (brendanheywood) [reply classes/metric/concurrent_users_metrics.php]: hi @marcghaly 

I have a few concerns around the public interface, and there is a lot of overlap with issue #73 which @jaypha was working on here:

https://github.com/catalyst/moodle-tool_cloudmetrics/pull/83

That pr isn't ready but what I've proposed just now with Jason is that the public interface for normal metrics is changed to something 

get_metric_item($starttime, $finishtime, $periodlength)

This is because in reality all metrics are retrospective, they just might be a few seconds later due to the latency of running each cron process and we need to know the exact timeperiod that the data should be have been for (even if the metric itself ignores it and just gets 'now')

Once that is in place the retrospective metrics should end up very similar something like:

get_metric_items($starttime, $finishtime, $periodlength)

The only difference is that the start time and finish time will be some distant past timestamps and we are expecting an array of them back instead of just one. We also need to consider the problem when the number of metric items returned is vast, and we might need to limit it and return what the time period was that we are returning data for which will be a subset of what was requested. Then it can be iteratively called to progressively backfill data in smaller chunks over time and the calling code would need to manage what has been processed or not already.

So I think we should land the first PR first so you don't have a moving target. Jason will get onto that in the next day or say.

I would also still like to meet, ideally all 3 of us if we can make it work, half as a meet and greet and also to iron out any last bits on this. I'll add a couple other comments in the pr here but many of them will be redundant when this gets rewritten on top of the other pr

File: classes/metric/concurrent_users_metrics.php [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            $interval = $secondsinterval[$data['interval']] ?? MINSECS * 5;
+            $sample = $sampleoptions[$data['sample']] ?? 10;
+        } else {
+            // If form is empty we do not want to backfill anything.
+            return null;
+        }
+
+        $sql = "SELECT floor(timecreated / $interval ) * $interval AS time,
+                       COUNT(DISTINCT userid) as value
+                  FROM {logstore_standard_log}";
```
Comment (brendanheywood) [inline classes/metric/concurrent_users_metrics.php]: We should not be querying directly from logstore_standard_log, there is an api for querying the logs instead:

https://github.com/moodle/moodle/blob/master/admin/tool/log/classes/log/manager.php#L93-L102

File: classes/metric/base.php [hunk 11]
```diff
@@ -131,4 +131,23 @@ public function get_settings_url() {
...
+        return false;
+    }
+
+    /**
+     * Returns records for backfilled metric.
+     *
+     * @param array $data Array of data specifying timeframe and sample to backfill.
+     * @return array|null
+     */
+    public function get_saved_metric_data(array $data): ?array {
```
Comment (brendanheywood) [inline classes/metric/base.php]: Can I suggest a name tweak like get_metric_items or get_metric_items_for_period?

I also dislike the $data argument, this would be much more readable as explicit arguments

File: classes/metric/concurrent_users_metrics.php (hunk 7 above)
Comment (marcghaly) [reply classes/metric/concurrent_users_metrics.php]: I have removed concurrent_users_metric class.

File: collector/database/classes/collector.php (hunk 9 above)
Comment (marcghaly) [reply collector/database/classes/collector.php]: Fixed.

File: classes/metric/concurrent_users_metrics.php (hunk 10 above)
Comment (marcghaly) [reply classes/metric/concurrent_users_metrics.php]: Hi @brendanheywood,

I have used the API, split recordset into chunks - but it seems using the API is consuming a lot of memory, any suggestion on this ?

https://github.com/catalyst/moodle-tool_cloudmetrics/blob/MOODLE_35_STABLE-67/classes/metric/online_users_metric.php#L122

File: classes/metric/base.php (hunk 11 above)
Comment (marcghaly) [reply classes/metric/base.php]: Fixed

**Comment by marcghaly** [comment]
Hi @brendanheywood,

I have made changes for this PR, a concern that I have is the use of the API for retrieving logs that does not provide a faster / lighter way to retrieve concurrent users (ie: using DISTINCT in a SELECT statement - retrieving all data `*` with the 2 methods provided by the API `get_events_select_iterator`, `get_events_select` seems super heavy to me, please let me know if there is a way I am not aware of) - it seems it is taking a lot of time for this operation, if that were to be left as is, would it need a spinner display ? This is rebased against latest MOODLE_35_STABLE. 
Concerning a meeting, I am available at 8 PM EST so 10 AM AEST your time, this evening and tomorrow if you want.

Regards,

Marc

**Comment by brendanheywood** [comment]
hi @marcghaly 

If the log api is too slow then do it directly like it was. But just to be be clear @jaypha was doing  / has just done a huge refactor of the way the metrics work and I didn't want to waste your time as it would affect your patch. I suspect you will need to have a look at what has landed and rewrite a lot of it (despite there being no conflicts)

We have our standup at 10, so lets chat right after that. Can you please ping me on matrix, I pinged you the other day but you didn't see it and I'm not sure I have the real 'you' in matrix

**Comment by marcghaly** [comment]
Hi @brendanheywood,

I have made the following changes:

- Used `metric_item class` directly in `generate_metric_items` instead of re-calling `generate_metric_item`.
- Added the UI on another page with one input representing period.
- Displayed what data can be retrieved - and what data has already been retrieved.
- Used the interval from the `Frequency` setting in `Manage Metrics` panel.
- Added a link to backfill support in the `Manage Metrics` panel.
- Queried the logstore directly for performance purposes, so it supports mariadb and pgsql.
- Added a renderer/mustache template for chart page.
- Instead of erasing data present in `cltr` db, when date is further than one already present, only complete data.
- If frequency is changed, data would be erased - it seems confusing having different intervals displayed - can change though.

Let me know if this suits.

Regards,

Marc

**Comment by brendanheywood** [comment]
hi @marcghaly 

This is looking great and pretty close, one issue I've found is that when you backfill data you don't end up with nice time periods which align with time period boundaries:

image

If we had a setting that was supposed to have been every hour then I'd expect the backfilled data to be on the hour every hour

A few other small comments in the pr

File: classes/admin_setting_manage_metrics.php [hunk 12]
```diff
@@ -155,7 +164,8 @@ public function output_html($data, $query='') {
                 $displayname,
                 $description,
                 $OUTPUT->render($editable),
-                $hideshow . $settingslink . $chartlink
+                $hideshow . $settingslink . $chartlink,
+                $backfilllink
```
Comment (brendanheywood) [inline classes/admin_setting_manage_metrics.php]: a super small general thing, with arrays it is nicer to add the trialing , on the last item so that next time we add an entry you don't get a diff which modifies an unrelated line

File: classes/collector/base.php [hunk 13]
```diff
@@ -56,4 +56,13 @@ public function record_metrics(array $metrics) {
     public function is_ready(): bool {
         return true;
     }
+
+    /**
+     * Abilitity for a collector to save old data.
+     *
+     * @return bool
+     */
+    public function supports_saved_metric_record(): bool {
```
Comment (brendanheywood) [inline classes/collector/base.php]: The method below is called is_backfillable, I think we should be consistent with the names in code and comments, eg something like:

is_backfillable()
supports_backfillable_metrics()

**Comment by jaypha** [comment]
Hi Marc

> If we had a setting that was supposed to have been every hour then I'd expect the backfilled data to be on the hour every hour

In case you didn't know, there are several functions in the classes/lib.php which can help with this.

File: classes/metric/base.php [hunk 14]
```diff
@@ -154,4 +177,37 @@ public function set_last_generate_time(int $timestamp) {
...
+
+    /**
+     * Returns two dates representing start end end log points, hence available data to retrieve.
+     *
+     * @return object|bool
+     */
+    public function get_range_log_available() {
+        global $DB;
+
+        $sql = 'SELECT max(timecreated) AS "max", min(timecreated) AS "min" FROM {logstore_standard_log}';
```
Comment (brendanheywood) [inline classes/metric/base.php]: would be slightly nicer to format this
https://docs.moodle.org/dev/SQL_coding_style#Indentation

File: classes/metric/online_users_metric.php [hunk 15]
```diff
@@ -37,4 +49,99 @@ class online_users_metric extends builtin_user_base {
     public function get_name(): string {
         return 'onlineusers';
     }
+
+    /**
+     * The metric type.
+     *
+     * @return int
+     */
+    public function get_type(): int {
```
Comment (brendanheywood) [inline classes/metric/online_users_metric.php]: Is this needed in this patch? Seems unrelated and isn't used elsewhere in the patch?

File: classes/metric/base.php [hunk 16]
```diff
@@ -154,4 +177,37 @@ public function set_last_generate_time(int $timestamp) {
...
+
+    /**
+     * Returns records for backfilled metric.
+     *
+     * @param int $backwardperiod Time from which sample is to be retrieved.
+     * @param int $finishtime If data is being completed argument is passed here.
+     *
+     * @return array|null
+     */
+    public function generate_metric_items(int $backwardperiod, int $finishtime = null): ?array {
```
Comment (brendanheywood) [inline classes/metric/base.php:199]: One thing I would like to clear up in the contract for a backfillable metric and the php docs etc is the alignment of dates for the period and finish time.

For instance, lets say you have a metric which is run every day, but we only started today. We want to back fill it for a year. Right now it is midday and we start by back-filling it for just a week. We need to be clear about what the starting point is for the period and weather this is 100% the responsibility of the metric to calculate correctly.

Because if we ask for 1 week of data, but going back from now at midday then we will end up with half a days data left over at the other end. We could end up saving an incorrect metric because its only counting part of the data for that period. The same is true of the $finishtime. I don't mind which side of the fence it is on as long as it works and is clearly documented, but it is probably better shared as much as possible and done on the collector chart side. ie $finishtime will always be exactly on a boundary of a frequency time slice, and $backwardperiod will always be some integer product of the frequency.

I also think there will be subtle bugs when it comes to daylights savings but I think we can just ignore it and live with the slight errors. At least all the data will appear in one metric or the adjacent one so it should mostly balance out.

File: classes/metric/online_users_metric.php [hunk 17]
```diff
@@ -37,4 +49,99 @@ class online_users_metric extends builtin_user_base {
...
+            manager::FREQ_12HOUR => HOURSECS * 12,
+            manager::FREQ_DAY => DAYSECS,
+            manager::FREQ_WEEK => WEEKSECS,
+            manager::FREQ_MONTH => WEEKSECS * 4
+        ];
+
+        $interval = $secondsinterval[$frequency];
+        $sql = 'SELECT floor(timecreated / :interval ) * :intervaldup AS time,
+                       COUNT(DISTINCT userid) as value
+                  FROM {logstore_standard_log} WHERE timecreated >= :starttime AND timecreated <= :finishtime
```
Comment (brendanheywood) [inline classes/metric/online_users_metric.php]: a few line breaks with sql indents here

File: classes/metric/online_users_metric.php [hunk 18]
```diff
@@ -37,4 +49,99 @@ class online_users_metric extends builtin_user_base {
...
+    }
+
+    /**
+     * Stores what data has been sent to collector.
+     *
+     */
+    public function set_data_sent_config() {
+        // Store what data has been sent min, max timestamp range and interval.
+        $currentconfig = $this->mintimestamp . '-' . $this->maxtimestamp . '-' . $this->interval;
+        $this->sameconfig = ($this->get_range_retrieved() === $currentconfig) ? true : false;
```
Comment (brendanheywood) [inline classes/metric/online_users_metric.php]: should not need the ternary at the end

File: collector/database/chart.php [hunk 19]
```diff
@@ -59,16 +65,57 @@
...
     $metriclabels,
     $metricname
 );
 $select->set_label(get_string('select_metric_for_display', 'cltr_database'));
+$context = [];
+// Error management if metric is not enabled.
+if (!isset($metrics[$metricname])) {
+    throw new moodle_exception('metric_not_enabled', 'tool_cloudmetrics', '', $metricname);
+}
+$backfilledmetric = $metrics[$metricname]->is_backfillable();
```
Comment (brendanheywood) [inline collector/database/chart.php]: This page is about showing the chart, I'd prefer it if all the logic around backfilling was moved to a new endpoint which you post to, and then after it has filled the data then it redirects back to the chart page as a get.

**Comment by marcghaly** [comment]
Hi @brendanheywood,

Just checking, in the following example it is set to be hourly, for example between two first rows, should it display between 12:00 and 16:00 (12, 13, 14, 15 , 16) with a 0 value ? Is that what you mean ? I am using pgsql maybe issue is related to mariadb/mysqli 

[](url)
Screenshot from 2022-06-07 09-39-17

This is what I get with mysql:
Screenshot from 2022-06-07 09-57-47


Regards,

Marc

File: collector/database/chart.php [hunk 20]
```diff
@@ -59,16 +65,57 @@
...
+    $daterange = $metrics[$metricname]->get_range_log_available();
+    $startdate = userdate($daterange->min, get_string('strftimedatetime', 'cltr_database'), $CFG->timezone);
+    $enddate = userdate($daterange->max, get_string('strftimedatetime', 'cltr_database'), $CFG->timezone);
+    $mform = new metric_backfill_form(null, $daterange);
+    $context['form'] = $mform->render();
+    if ($fromform = $mform->get_data()) {
+        $periodretrieval = $fromform->periodretrieval;
+        $finishtime = $datametricsaved[0] ?? null;
+        $metricitems = $metrics[$metricname]->generate_metric_items($periodretrieval, $finishtime);
+        if ($collector->supports_saved_metric_record()) {
```
Comment (brendanheywood) [inline collector/database/chart.php]: I think check this much earlier at the top and throw an error if it is not backfillable

File: classes/metric/base.php [hunk 21]
```diff
@@ -125,6 +125,29 @@ public function get_settings_url() {
         return null;
     }
 
+    /**
+     * The range for which a backfilled metric has retrieved data.
+     *
+     * @return string|bool
+     * @throws \dml_exception
+     */
+    public function get_range_retrieved() {
```
Comment (brendanheywood) [inline classes/metric/base.php:134]: would it be nicer to the calling code if this returned an array of two values? would remove a lot of explodes around the place

File: collector/database/classes/collector.php [hunk 22]
```diff
@@ -72,4 +88,40 @@ public function get_metrics($metricnames = null, $since = false): array {
...
+     * @param  \tool_cloudmetrics\metric\base $metricclass Class representing metric.
+     * @param  array $data Array of metric items.
+     */
+    public function record_saved_metrics(\tool_cloudmetrics\metric\base $metricclass, array $metricitems = []) {
+        global $DB;
+
+        $maxbackdate = explode('-', $metricclass->get_range_retrieved())[0];
+        $transaction = $DB->start_delegated_transaction();
+        // Sets what data has been sent to collector.
+        $metricclass->set_data_sent_config();
```
Comment (brendanheywood) [inline collector/database/classes/collector.php]: you should avoid setting config inside a transaction, if it rolls back you can end up with the db and the MUC out of sync because the MUC doesn't roll back. I'd do it after it has committed

File: collector/database/classes/collector.php [hunk 23]
```diff
@@ -72,4 +88,40 @@ public function get_metrics($metricnames = null, $since = false): array {
...
+        if (is_array($metricitems) && count($metricitems) != 0 && !$metricclass->sameconfig) {
+            // When a further date is passed we only want to add missing data.
+            $iscompleted = ($metricitems[0]->time < $maxbackdate) ? true : false;
```
Comment (brendanheywood) [inline collector/database/classes/collector.php]: again ternary is redundant

File: collector/database/classes/collector.php [hunk 24]
```diff
@@ -72,4 +88,40 @@ public function get_metrics($metricnames = null, $since = false): array {
...
+            if ($iscompleted) {
+                // Trick to avoid erasing data when it only needs to be completed.
+                $this->record_metrics($metricitems);
+            } else {
+                $this->delete_metrics($metricclass->get_name());
```
Comment (brendanheywood) [inline collector/database/classes/collector.php]: I'm a little confused why we'd ever need to delete data? If we are being asked to backfill data from time period which already has data then that smells like what ever is calling record_saved_metrics isn't sending correct times

File: collector/database/classes/form/metric_backfill_form.php [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+ * @author    Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class metric_backfill_form extends moodleform {
+    public function definition() {
+        $mform = $this->_form;
+        $maxbackwardsdate = $this->_customdata->min;
+        $maxperiod = time() - $maxbackwardsdate;
+        $periods = [
+            HOURSECS      => get_string('one_hour', 'tool_cloudmetrics'),
```
Comment (brendanheywood) [inline collector/database/classes/form/metric_backfill_form.php]: Can this be refactored to dedup the same array at:

https://github.com/catalyst/moodle-tool_cloudmetrics/pull/92/files#diff-0b6a2a0fb28e2d23493ed5518e61238fdcae2431af1184bc1b8b1578fda8a007R120

**Comment by brendanheywood** [comment]
> Just checking, in the following example it is set to be hourly, for example between two first rows, should it display between 12:00 and 16:00 (12, 13, 14, 15 , 16) with a 0 value ? Is that what you mean ? I am using pgsql maybe issue is related to mariadb/mysqli

The golden rule is imagine that the metric had been running the whole time and we didn't need to back fill. Ideally the end result should be exactly the same. So yes there would be a metric for every time slice and many of them would be 0

**Comment by brendanheywood** [comment]
Also I am using postgress too

File: classes/admin_setting_manage_metrics.php (hunk 12 above)
Comment (marcghaly) [reply classes/admin_setting_manage_metrics.php]: Added

File: classes/collector/base.php (hunk 13 above)
Comment (marcghaly) [reply classes/collector/base.php]: Fixed.

File: classes/metric/base.php (hunk 14 above)
Comment (marcghaly) [reply classes/metric/base.php]: Fixed

File: classes/metric/online_users_metric.php (hunk 15 above)
Comment (marcghaly) [reply classes/metric/online_users_metric.php]: No it is present in parent class, removed it.

File: classes/metric/base.php (hunk 16 above)
Comment (marcghaly) [reply classes/metric/base.php:199]: Would you like to have plain days instead of sliced interval data ? I will document and add changes based on your answer.

File: classes/metric/online_users_metric.php (hunk 17 above)
Comment (marcghaly) [reply classes/metric/online_users_metric.php]: Fixed

File: classes/metric/online_users_metric.php (hunk 18 above)
Comment (marcghaly) [reply classes/metric/online_users_metric.php]: Fixed

File: collector/database/classes/collector.php (hunk 22 above)
Comment (marcghaly) [reply collector/database/classes/collector.php]: Fixed.

File: classes/metric/base.php (hunk 21 above)
Comment (marcghaly) [reply classes/metric/base.php:134]: Fixed

File: collector/database/classes/collector.php (hunk 23 above)
Comment (marcghaly) [reply collector/database/classes/collector.php]: Fixed.

File: collector/database/classes/collector.php (hunk 24 above)
Comment (marcghaly) [reply collector/database/classes/collector.php]: I found it confusing seeing different frequencies in the chart - what happens at the moment : it retrieves all data for the requested time-frame unless the frequency changed, when frequency changes, it deletes all data and replaces it with the new data with proper interval. Also, if I request data for 4 months, it will fill data for those 4 months then if I request data for 12 months, it will retrieve data from the lowest timestamp present in DB to the passed point time requested (with the period) - it will only retrieve the missing part.

File: collector/database/classes/form/metric_backfill_form.php (hunk 25 above)
Comment (marcghaly) [reply collector/database/classes/form/metric_backfill_form.php]: I have passed it as part of the 2nd parameter of the moodle form.

File: collector/database/chart.php (hunk 19 above)
Comment (marcghaly) [reply collector/database/chart.php]: I have split those in two pages chart + backfill

File: collector/database/chart.php (hunk 20 above)
Comment (marcghaly) [reply collector/database/chart.php]: I have added an exception earlier to the `backfill` page.

**Comment by marcghaly** [comment]
Hi @brendanheywood,

I have made changes according to your comments, If changes are okay I will squash commits.

Cheers,

Marc

File: classes/metric/base.php [hunk 26]
```diff
@@ -125,6 +125,29 @@ public function get_settings_url() {
...
+        }
+        return [$config[0], $config[1], $config[2]];
+    }
+
+    /**
+     * Sets the range for which a backfilled metric has retrieved data.
+     *
+     * @param string $currentconfig Config containing min and max timestamps plus interval.
+     */
+    public function set_range_retrieved(string $currentconfig) {
```
Comment (brendanheywood) [inline classes/metric/base.php]: would it now make sense for set and get to take the same shaped data,ie array of 3 things

File: classes/metric/base.php (hunk 16 above)
Comment (brendanheywood) [reply classes/metric/base.php:199]: Not quite sure what you mean by plain days. To me if the metric happens to be once a day then it would be time slices of each day and each day starts at midnight and ends at midnight. But this all depends on the metric frequency, the same applies if it is an hourly metric or every 5 minutes

File: collector/database/classes/collector.php (hunk 24 above)
Comment (brendanheywood) [reply collector/database/classes/collector.php]: I think deleting data has some business value but I think it should be explicit, similar to the setting for how much data you want to keep around. I don't like the idea of someone trying to add in more data and ending up losing data they already had.

I've split that off into a new issue, if you want you can tackle that after this pr lands:
https://github.com/catalyst/moodle-tool_cloudmetrics/issues/104

For this one I'd just want it to only ever add more data

File: classes/metric/base.php (hunk 26 above)
Comment (marcghaly) [reply classes/metric/base.php]: Fixed.

File: collector/database/classes/collector.php (hunk 24 above)
Comment (marcghaly) [reply collector/database/classes/collector.php]: I have set it up so it never deletes data now - but also that if frequency changed it is not overlapping data of another interval - It would get extremely confusing. If frequency changes, timestamp needs to be earlier than lowest one in DB so it completes data, it looks like this. Highlighted the one month period - the other one 1 week.

Screenshot from 2022-06-08 15-54-27

I will sure get on this one once this PR lands.

**Comment by marcghaly** [comment]
Hi @brendanheywood, this is squashed and rebased over latest master.

Regards,

Marc

**Comment by brendanheywood** [comment]
thanks @marcghaly
